text/0000-block-manager.md

```diff
@@ -0,0 +1,367 @@
+3. For any two sequential blocks in the list, the second block's parent is
+   the first block
+
+If any of these three conditions are not met, an exception is raised.
+
+All blocks added will have a reference count of 1 if the operation is
+successful, including the tip.
+
+`get(block_ids: List<String>) -> Iter<Block>`
+
+    Atomically get the blocks with the given ids. This passes through to any
+    registered stores as necessary.
+
+`branch(tip: String) -> Iter<Block>`
+
+    Create an iterator over blocks on a given branch starting with the block
+    with the id given in `tip` and traversing from the tip to the root.
+
+`branch_diff(tip: String, exclude: String) -> Iter<Block>`
+
+    Create an iterator over blocks on a given branch starting with the block
+    with the id given in `tip` and traversing from the tip to the root,
+    excluding blocks that are also in the branch starting with id in exclude.
+
+### Holding Blocks
+
+In order to avoid removing blocks that are in the process of being worked on,
+the block manager provides an operation to place a `hold` on a block. When this
+operation is called, the block manager makes a guarantee that the block will
+not be dropped until corresponding `drop` is called. Consequently, when an
+object that requests a hold is done with the block, it must call `drop` to
+signal it is done with the block.
+
+Ensuring that blocks depended on by external objects are not removed is handled
+by the following methods:
+
+`ref_block(block_id: String)`
+
+    Ensure that the block with the given block id is not dropped until a
+    corresponding `unref_block()` with the same block id is called.
+
+`unref_block(block_id: String)`
+
+    Release a previous `ref_block()` on a block. May result in the block being
+    removed from the tree.
+
+    Raises an exception if the reference count falls below 0, which indicates
+    a bug in the application code.
+
+The following example illustrates how the `ref_block` operation is used to
+prevent blocks from being dropped when they are transferred from the completer
+to the chain controller.
+
+Assume that the chain controller has a block A and that the completer has just
+completed a block B and would like to pass it to the chain controller. If the
+completer were to just pass the chain controller B through a queue or shared
+memory, the chain controller could decide to `unref_block` A before it receives
+B, causing B to have a missing predecessor when it arrives at the chain
+controller. Instead, the completer first places a `ref_block` on A, causing and
+then it passes B to the chain controller. If the chain controller decides to
+`unref_block` A now, the block manage will ensure it is not actually dropped,
+since there is still an open hold on the block. Finally, when the chain
+controller takes ownership of B, it can send a signal back to the completer
+that it has been received and the completer can release its hold.
```

Comment:
"completer can release its hold": Does the completer have a hold on the block or is it the Block Manager?

Reply:
And what does it mean for the chain controller to take ownership of B? Would the completer actually be passing a reference to the controller for an object managed by the block manager?